src/logs/mod.rs

```diff
@@ -17,54 +17,61 @@ declare_log_buffer!(name = DEBUG_BUF, capacity = 1000);
 // Trace of HTTP requests and responses.
 declare_log_buffer!(name = TRACE_HTTP_BUF, capacity = 1000);
 
-pub const INFO: PrintProxySink = PrintProxySink(&LogMessageType::Info, &INFO_BUF);
-pub const DEBUG: PrintProxySink = PrintProxySink(&LogMessageType::Debug, &DEBUG_BUF);
-pub const TRACE_HTTP: PrintProxySink = PrintProxySink(&LogMessageType::TraceHttp, &TRACE_HTTP_BUF);
+pub const INFO: PrintProxySink = PrintProxySink(&Priority::Info, &INFO_BUF);
+pub const DEBUG: PrintProxySink = PrintProxySink(&Priority::Debug, &DEBUG_BUF);
+pub const TRACE_HTTP: PrintProxySink = PrintProxySink(&Priority::TraceHttp, &TRACE_HTTP_BUF);
+
+#[cfg(test)]
+thread_local! {
+    pub static DISPLAYED_LOG_ENTRIES: std::cell::RefCell<Vec<ic_canister_log::LogEntry>> = std::cell::RefCell::new(vec![]);
+}
 
 #[derive(Debug)]
-pub struct PrintProxySink(&'static LogMessageType, &'static GlobalBuffer);
+pub struct PrintProxySink(&'static Priority, &'static GlobalBuffer);
 
 impl Sink for PrintProxySink {
     fn append(&self, entry: ic_canister_log::LogEntry) {
-        let message_type = self.0;
-        if get_log_message_filter().should_print_log_message(*message_type) {
-            ic_cdk::println!(
-                "{} {}:{} {}",
-                message_type.as_str_uppercase(),
-                entry.file,
-                entry.line,
-                entry.message,
-            )
+        let message = format!(
+            "{} {}:{} {}",
+            self.0.as_str_uppercase(),
+            entry.file,
+            entry.line,
+            entry.message,
+        );
+        if get_console_filter().is_match(&message) {
+            ic_cdk::println!("{}", message);
+            #[cfg(test)]
+            DISPLAYED_LOG_ENTRIES.with_borrow_mut(|entries| entries.push(entry.clone()));
```

Comment:
mixing test code with productive code is usually not a good idea and I think here you actually don't need it, so I would remove this and the declaration L24-27.

Here is an alternative of how it could look like in the test file `src/logs/tests.rs`:
```rust
...

declare_log_buffer!(name = TEST_BUF, capacity = 1000);
const TEST: PrintProxySink = PrintProxySink(&Priority::Debug, &TEST_BUF);

...

#[test]
fn should_show_all() {
    set_console_filter(ConsoleFilter::ShowAll);
    log!(TEST, "ABC");
    log!(TEST, "123");
    log!(TEST, "!@#");
    assert_eq!(get_messages(), vec!["ABC", "123", "!@#"]);
}

...

fn get_messages() -> Vec<String> {
    export(&TEST_BUF)
        .into_iter()
        .map(|entry| entry.message)
        .collect()
}

```

Does-it make sense?

Reply:
Thanks for pointing this out as a much cleaner way to achieve this!

Reply:
The problem I'm running into with this approach is that the messages are always added to the buffer regardless of the filter, which only affects whether the messages are printed to the console.

Reply:
> the messages are always added to the buffer regardless of the filter, which only affects whether the messages are printed to the console.

good point, I missed this. I should have tried with more tests :)

> The simplest solution would be to change the logic so that the console messages are the same as the logs.

I have the impression this would indeed the simplest. I guess for this it would then maybe make sense to rename `ConsoleFilter` to something like `LogFilter`?

Reply:
Done!